autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/AutofillServiceLifecycleObserver.kt

```diff
@@ -0,0 +1,97 @@
+import android.content.pm.PackageManager.COMPONENT_ENABLED_STATE_DISABLED_USER
+import android.content.pm.PackageManager.COMPONENT_ENABLED_STATE_ENABLED
+import android.content.pm.PackageManager.DONT_KILL_APP
+import androidx.lifecycle.LifecycleOwner
+import com.duckduckgo.app.di.AppCoroutineScope
+import com.duckduckgo.app.lifecycle.MainProcessLifecycleObserver
+import com.duckduckgo.common.utils.DispatcherProvider
+import com.duckduckgo.di.scopes.AppScope
+import com.squareup.anvil.annotations.ContributesMultibinding
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import timber.log.Timber
+
+@ContributesMultibinding(
+    scope = AppScope::class,
+    boundType = MainProcessLifecycleObserver::class,
+)
+class AutofillServiceLifecycleObserver @Inject constructor(
+    @AppCoroutineScope private val appCoroutineScope: CoroutineScope,
+    private val dispatcherProvider: DispatcherProvider,
+    private val context: Context,
+    private val autofillServiceFeature: AutofillServiceFeature,
+) : MainProcessLifecycleObserver {
+
+    override fun onCreate(owner: LifecycleOwner) {
+        super.onCreate(owner)
+        appCoroutineScope.launch(dispatcherProvider.io()) {
+            runCatching {
+                val currentState = getAutofillServiceState(context).toBoolean()
+                autofillServiceFeature.self().isEnabled().let { remoteState ->
+                    if (currentState != remoteState) {
+                        Timber.d("DDGAutofillService: Updating state to $remoteState")
+                        newState(context, remoteState)
+                    }
+                }
+            }.onFailure {
+                Timber.e("DDGAutofillService: Failed to update Service state: $it")
+            }
+        }
+    }
+
+    private fun getAutofillServiceState(context: Context): Int {
+        val pm = context.packageManager
+        val autofillServiceComponent = ComponentName(context, RealAutofillService::class.java)
+        return pm.getComponentEnabledSetting(autofillServiceComponent)
+    }
+
+    private fun newState(
+        context: Context,
+        isEnabled: Boolean,
+    ) {
+        val pm = context.packageManager
+        val autofillServiceComponent = ComponentName(context, RealAutofillService::class.java)
+
+        val value = when (isEnabled) {
+            true -> COMPONENT_ENABLED_STATE_ENABLED
+            false -> COMPONENT_ENABLED_STATE_DISABLED
+        }
+
+        pm.setComponentEnabledSetting(autofillServiceComponent, value, DONT_KILL_APP)
+    }
+
+    private fun Int.toBoolean(): Boolean? {
+        return when (this) {
+            COMPONENT_ENABLED_STATE_DEFAULT -> false // this is the current value in Manifest
+            COMPONENT_ENABLED_STATE_ENABLED -> true
+            COMPONENT_ENABLED_STATE_DISABLED -> false
+            COMPONENT_ENABLED_STATE_DISABLED_USER -> true
+            COMPONENT_ENABLED_STATE_DISABLED_UNTIL_USED -> true
```

Comment:
these should return `false` ?

Reply:
Discussed this offline. For those 2 values we've decided to take no action. They seem to be external to us, one even disabled by the user, so we should honor those disabled states and don't enable our service by our remote config.

New approach returns null for those states, and we don't update the value.